### DISCUSSION TITLE
The Row Height Conondrum

### DISCUSSION THREAD

**Discussion by lordofscripts** [johnfercher/maroto#458] [Ideas]
I think that burdening the end-user (users of the library) with having to specify a **row height** should actually be addressed in version 2 (latest release is 2.0.7) rather than 3. Currently the library works well for form-making rather than document-making.

I believe `Maroto/V2` should address these 3 low-level) issues and not burden the end-developer with having to deal with these as it is currently (`v2.0.7`).

**1st Problem: Row Height**

There is a _certain_ relation between **font height** and **row height** which is applicable in _single line per row_  use cases. 

In this case, in a first attempt, with curve fitting (regression), _a simple cubic (3rd order polynomial) provides a good prediction of the single-line row height necessary for a given font height_.

**2nd Problem: How many rows in a page**

My experiments revealed that for a **Letter-sized** page without header and a single-line footer (page nr.), a single page fits **240 rows**, for example 60 rows of a row height of 4.

Obviously, that amount is slightly different for an A4 page.

**3rd Problem: Multi-line rows**

This issue is of **extreme** importance for using Maroto in regular document generation where text is composed of paragraphs (multiple lines).

In this situation we have several variables to predict the necessary row height for a given text of 1..n paragraphs:

- **Font Height**: bigger font, less character space available in a single line. _Inversely proportional, big influence l_.
- **Font Family**: some fonts (narrow) occupy less horizontal space than others. _Medium influence_.
- **Font Style**: Bold & Italic have a slight influence on how many characters fit in a single line.
- **Horizontal space available**: The column span in Maroto which is typically 1..12.

With that in mind there are methods to compute font metrics, knowing how a paragraph of a given font size, family & style would be split into several lines. Given those parameters and a mapping function, we could know how many row height units are needed to properly output a (series of) paragraphs.

Y: number of lines
X: Font metrics (family, size, style)
Z: Column Span (horizontal space)

`y = f(x, z)`

**Comment by lordofscripts** [comment]
May be of help: [Go MeasureString](https://pkg.go.dev/golang.org/x/image/font) provides `font.MeasureString()`. 

But at this moment the documentation here does not clarify the **font size unit**__, is it _points, millimeters, pixels_?

**Comment by lordofscripts** [comment]
Additionally, I noticed that `Maroto` has the GO `pdfcpu` library as an indirect dependency. It so happens that `pdfcpu` has a lot of built-in functionality to calculate these things we `Maroto` users need (and miss), and that it could be exposed by Maroto.

**Comment by Fernando-hub527** [comment]
I don't know if this is the right topic to discuss this issue, but...

I am implementing this feature to automatically calculate the line size, but I realized that the matrixcode component has its size calculated based on the available space (height and width). In this case I see two options:

- Calculate the size of the component only based on the available width and, based on the result, find the height necessary to accommodate the component.
- replace the percentage property with a fixed value, requiring the user to directly inform the size he expects the component to have and then find the required height.

I'm really unsure of the best way
